daemon/main.cpp

```diff
@@ -835,7 +835,12 @@ bool Daemon::maybe_stats(bool send_ping)
 
 #endif
 
-        // Matz got in the urine that not all CPUs are always feed
+        // Heuristic for the memory limit on compile jobs:
+        //   1) If we have four or more compile slots, allow each job 25% of the available (physical) memory.
+        //   2) If we have three compile slots, allow each job 1/3 of the available (physical) memory.
+        //   3) If we have two compile slots, allow each job 1/2 of the available (physical) memory.
+        //   4) If we have one compile slot, allow it to use all of the available (physical) memory.
+        //   5) Allow each job at least `min_mem_limit` megabytes.
         mem_limit = std::max(int(msg.freeMem / std::min(std::max(max_kids, 1U), 4U)), min_mem_limit);
```

Comment:
Hmm, is that one line really that cryptic that it needs 6 lines of description? The original comment is indeed rubbish, but this looks like overdoing it way too much in the opposite direction.

Reply:
I expanded this to hopefully get some confirmation that it works as I think it does, before I try to tweak the heuristic (we have started to hit this limit more often after a recent compiler change).  Is it worth shortening this a bit to something like: 

> 1) If we have between one and three compile slots, allow each job 1/n of the available (physical) memory.
> 2) If we have more compile slots, allow each job 1/4 of the available physical memory.
> 3) Allow each job at least `min_mem_limit` megabytes.

Possible changes that I have in mind:
* divide the available memory by the number of unused compile slots (instead of the total number of compile slots, as it seems to be now)
* increase the memory limit if swap is available
* make min_mem_limit a daemon config option
* allow the client to specify a suitable memory limit (or minimum memory limit), but this might be more involved, possibly requires a protocol version bump